### PR TITLE
Integration of vstest.console process with datacollector process.

### PR DESCRIPTION
**Pull request by harshjain2**
Integration of vstest.console process with datacollector process.

Related to :
https://github.com/Microsoft/vstest/issues/309

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs [hunk 1]
```diff
@@ -291,7 +292,7 @@ public void HandleTestRunComplete(TestRunCompleteEventArgs runCompleteArgs, Test
                             runCompleteArgs.IsCanceled,
                             runCompleteArgs.IsAborted,
                             runCompleteArgs.Error,
-                            null,
+                            new Collection<AttachmentSet>(new List<AttachmentSet>(runContextAttachments)),
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: Why not pass `runContextAttachments` directly?

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs [hunk 2]
```diff
@@ -363,7 +364,7 @@ public virtual void HandleTestRunStatsChange(TestRunChangedEventArgs testRunChan
                 EqtTrace.Info("TestRunRequest:SendTestRunStatsChange: Completed.");
             }
         }
-        
+
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: nit: remove blank line

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs [hunk 3]
```diff
@@ -100,10 +107,23 @@ public void Initialize(SessionId id, string outputDirectory, IMessageSink messag
         }
 
         /// <inheritdoc/>
-        public List<AttachmentSet> GetAttachments(DataCollectionContext dataCollectionContext)
+        public List<AttachmentSet> GetAttachments(DataCollectionContext dataCollectionContext, bool isCancelled = false)
         {
-            Task.WhenAll(this.attachmentTasks.ToArray()).Wait();
-            return this.AttachmentSets.Values.ToList();
+            if (isCancelled)
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs]: Should cancel be a responsibility of GetAttachments?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorLoader.cs [hunk 4]
```diff
@@ -30,6 +29,7 @@ internal class DataCollectorLoader : IDataCollectorLoader
 #else
                 assembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(location);
 #endif
+                var dataCollectorType = assembly.GetTypes().ToList().Where(t => t.AssemblyQualifiedName.Equals(assemblyQualifiedName)).FirstOrDefault();
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorLoader.cs]: Please remove ToList.

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs [hunk 5]
```diff
@@ -100,10 +107,23 @@ public void Initialize(SessionId id, string outputDirectory, IMessageSink messag
...
+            try
+            {
+                Task.WhenAll(this.attachmentTasks.ToArray()).Wait();
+            }
+            catch (Exception ex)
+            {
+                EqtTrace.Error(ex.Message);
+            }
+
+            return this.AttachmentSets?.Values.ToList();
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs]: When will AttachmentSets be null?

File: src/Microsoft.TestPlatform.Common/DataCollection/Interfaces/IDataCollectionManager.cs [hunk 6]
```diff
@@ -45,7 +45,12 @@ internal interface IDataCollectionManager : IDisposable
         /// <summary>
         /// Raises SessionEnd event to all data collectors configured for run.
         /// </summary>
-        /// <returns>Collection of session attachmentSet.</returns>
-        Collection<AttachmentSet> SessionEnded();
+        /// <param name="isCancelled">
+        /// Boolean to specify is the test ruun is cancelled or not.
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/Interfaces/IDataCollectionManager.cs]: nit: s/ruun/run

File: src/Microsoft.TestPlatform.Common/Friends.cs [hunk 7]
```diff
@@ -6,11 +6,16 @@
 #region Product Assemblies
 [assembly: InternalsVisibleTo("Microsoft.TestPlatform.CrossPlatEngine, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
 [assembly: InternalsVisibleTo("vstest.console, PublicKey =002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
+[assembly: InternalsVisibleTo("datacollector, PublicKey =002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Friends.cs]: Why are internalsvisibleto required for these?

File: src/Microsoft.TestPlatform.Common/Friends.cs [hunk 8]
```diff
@@ -6,11 +6,16 @@
...
+[assembly: InternalsVisibleTo("Microsoft.TestPlatform.CommunicationUtilities, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Friends.cs]: Not sure why TP.Common needs a dependency on CommunicationUtilities.

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs (hunk 1 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: done.

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorLoader.cs (hunk 4 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorLoader.cs]: done.

File: src/Microsoft.TestPlatform.Common/DataCollection/Interfaces/IDataCollectionManager.cs (hunk 6 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/DataCollection/Interfaces/IDataCollectionManager.cs]: done.

File: src/Microsoft.TestPlatform.Common/Friends.cs (hunk 7 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/Friends.cs]: For exposing IMessageSink interface that is in common, datacollector implements IMessageSink.

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs (hunk 1 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: done.

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs (hunk 2 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Microsoft.TestPlatform.CommunicationUtilities.csproj [hunk 9]
```diff
@@ -33,6 +33,11 @@
     <Reference Include="Microsoft.CSharp" />
   </ItemGroup>
   <ItemGroup>
+    <Reference Include="System.Dynamic.Runtime">
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Microsoft.TestPlatform.CommunicationUtilities.csproj]: Why is this change required?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs [hunk 10]
```diff
@@ -65,33 +58,18 @@ public void Initialize(Architecture architecture)
         public virtual int LaunchDataCollector(IDictionary<string, string> environmentVariables, IList<string> commandLineArguments)
         {
             var currentWorkingDirectory = Path.GetDirectoryName(typeof(DataCollectionLauncher).GetTypeInfo().Assembly.Location);
-            string dataCollectorProcessPath, processWorkingDirectory = null;
+            string dataCollectorProcessPath = null, processWorkingDirectory = null;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: How are we handling scenarios where process doesn't start? Or process crashes?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+            {
+                EqtTrace.Verbose("DotnetDataCollectionLauncher: Full path of dotnet.exe is {0}", currentProcessFileName);
+            }
+
+            var dataCollectorAssemblyPath = Path.Combine(currentWorkingDirectory, DataCollectorProcessName);
+            commandLineArguments.Insert(0, dataCollectorAssemblyPath);
+
+            dataCollectorFileName = Path.GetFileNameWithoutExtension(dataCollectorAssemblyPath);
+
+            // .NET core host manager is not a shared host. It will expect a single test source to be provided.
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: This is not true. Datacollector is always shared?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs [hunk 12]
```diff
@@ -20,13 +19,8 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.DataCollection
     internal class DataCollectionLauncher : IDataCollectionLauncher
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: Please follow the convention:
* DefaultDataCollectionLauncher
* DotnetDataCollectionLauncher

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+            }
+            else
+            {
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("DotnetDataCollectionLauncher: File {0}, doesnot exist", depsFilePath);
+                }
+            }
+
+            commandLineArguments.Insert(0, args);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: IList.Insert is O(n). We're doing it multiple times in this logic. Why not use String.Concat and String.Join(commandLineArguments)?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+                separator = ':';
+                dotnetExeName = "dotnet";
+            }
+#endif
+
+            var pathString = Environment.GetEnvironmentVariable("PATH");
+            foreach (string path in pathString.Split(separator))
+            {
+                var exeFullPath = Path.Combine(path.Trim(), dotnetExeName);
+                if (File.Exists(exeFullPath))
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: How are we testing this logic? We have to use IFileHelper here.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+            this.DataCollectorProcess = this.processHelper.LaunchProcess(currentProcessFileName, argumentsString, currentWorkingDirectory);
+            return this.DataCollectorProcess.Id;
+        }
+
+        /// <summary>
+        /// Get full path for the .net host
+        /// </summary>
+        /// <returns>Full path to <c>dotnet</c> executable</returns>
+        /// <remarks>Debuggers require the full path of executable to launch it.</remarks>
+        private string GetDotnetHostFullPath()
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: This functionality is DRY violation with DotnetTestHostManager. Please extract it to an independent unit.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs [hunk 16]
```diff
@@ -33,14 +33,20 @@ internal class ProxyDataCollectionManager : IProxyDataCollectionManager
...
-            : this(arch, settingsXml, new DataCollectionRequestSender(), new DataCollectionLauncher())
+        public ProxyDataCollectionManager(Architecture arch, string settingsXml, string targetFramework)
+            : this(arch, settingsXml, new DataCollectionRequestSender(), DataCollectionLauncherFactory.GetDataCollectorLauncher(targetFramework))
         {
         }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ProxyDataCollectionManager"/> class.
         /// </summary>
+        /// <param name="arch">
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: Boilerplate docs. Use `inheritdocs` if possible.

File: test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+            proxyDataCollectionManager.BeforeTestRunStart(true, true, this.mockTestMessageEventHandler.Object);
+
+            var attachments = proxyDataCollectionManager.AfterTestRunEnd(false, this.mockTestMessageEventHandler.Object);
+
+            Assert.AreEqual("CustomDataCollector", attachments[0].DisplayName);
+            Assert.AreEqual("my://custom/datacollector", attachments[0].Uri.ToString());
+            Assert.IsTrue(attachments[0].Attachments[0].Uri.ToString().Contains("filename.txt"));
+
+            proxyDataCollectionManager.Dispose();
```
Comment (codito) [inline test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs]: using () { ... } in all TestMethods.

File: test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        [TestMethod]
+        public void AfterTestRunShouldHandleSocketFailureGracefully()
+        {
+            var socketCommManager = new SocketCommunicationManager();
+            var dataCollectionRequestSender = new DataCollectionRequestSender(socketCommManager, JsonDataSerializer.Instance);
+#if NET46
```
Comment (codito) [inline test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs]: Why not use `this.dataCollectionLauncher`?

File: test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+            Assert.IsNotNull(result);
+
+            socketCommManager.StopClient();
+
+            var attachments = proxyDataCollectionManager.AfterTestRunEnd(false, this.mockTestMessageEventHandler.Object);
+
+            Assert.IsNull(attachments);
+
+            // Give time to datacollector process to exit.
+            Thread.Sleep(500);
```
Comment (codito) [inline test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs]: Please use `result.WaitForExit()`

File: test/datacollector.PlatformTests/datacollector.PlatformTests.csproj [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+      <Version>4.0.1</Version>
+    </PackageReference>
+  </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
+  </ItemGroup>
+  <Import Project="$(TestPlatformRoot)scripts\build\TestPlatform.targets" />
+
+  <Target Name="AfterBuild" Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">
+    <Copy
```
Comment (codito) [inline test/datacollector.PlatformTests/datacollector.PlatformTests.csproj]: Why is this required? Isn't the datacollector launched from `artifacts\Debug\netXX\...` directory in these tests?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs [hunk 21]
```diff
@@ -207,80 +227,90 @@ private void AddNewFileTransfer(FileTransferInformation fileTransferInfo, AsyncC
...
+                                EqtTrace.Error(
+                                    "DataCollectionAttachmentManager.TriggerCallBack: Error occurred while raising the file transfer completed callback for {0}. Error: {1}",
+                                    localFilePath,
+                                    e.ToString());
+                            }
+                        }
+                    },
+                this.cancellationTokenSource.Token);
 
             this.attachmentTasks.Add(task);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs]: Just monitoring the continuationTask will suffice?

**Review by codito — APPROVED** [review]
Looks good with few suggestions.

**Comment by codito** [comment]
For the resx changes, please run `build.cmd -xlf` and include xlf files changes in PR.

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs (hunk 2 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: done.

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs (hunk 3 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs]: Implemented a new method Cancel that will take care of cancellation.

File: src/Microsoft.TestPlatform.Common/DataCollection/Interfaces/IDataCollectionManager.cs (hunk 6 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/DataCollection/Interfaces/IDataCollectionManager.cs]: done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs (hunk 12 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: done.

File: test/datacollector.PlatformTests/datacollector.PlatformTests.csproj (hunk 20 above)
Comment (harshjain2) [inline test/datacollector.PlatformTests/datacollector.PlatformTests.csproj]: These tests are executed from `test\datacollector.PlatformTests\bin\Debug`, therefore, we need to manually copy these files.

File: test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs (hunk 19 above)
Comment (harshjain2) [inline test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs]: done.

File: test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs (hunk 18 above)
Comment (harshjain2) [inline test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs]: IDataCollectorLauncher doesn't expose Process which I am using to monitor the process.
I don't see a need to expose the process in interface, therefore I am using specific instances here.

File: test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs (hunk 17 above)
Comment (harshjain2) [inline test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs]: done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs (hunk 15 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: Done. Extracted an interface IDotnetHostHelper.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs (hunk 13 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs (hunk 14 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs]: done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs (hunk 10 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: This is handled in ProxyExecutionManagerWithDataCollection, the class responsible for initiating test execution when data collection is enabled.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Microsoft.TestPlatform.CommunicationUtilities.csproj (hunk 9 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Microsoft.TestPlatform.CommunicationUtilities.csproj]: not required, removed.
